pkg/utils/utils_mlx.go

```diff
@@ -0,0 +1,102 @@
+package utils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+//BlueField2 mode representation
+type BlueField2Mode int
+
+const (
+	bluefield2Dpu BlueField2Mode = iota
+	bluefield2Conntexd
+
+	internalCPUPageSupplier   = "INTERNAL_CPU_PAGE_SUPPLIER"
+	internalCPUEswitchManager = "INTERNAL_CPU_ESWITCH_MANAGER"
+	internalCPUIbVporto       = "INTERNAL_CPU_IB_VPORT0"
+	internalCPUOffloadEngine  = "INTERNAL_CPU_OFFLOAD_ENGINE"
+
+	ecpf      = "ECPF"
+	extHostPf = "EXT_HOST_PF"
+
+	disabled = "DISABLED"
+	enabled  = "ENABLED"
+)
+
+func MstConfigReadData(pciAddress string) (string, error) {
+	glog.Infof("MstConfigReadData(): device %s", pciAddress)
+	args := []string{"-e", "-d", pciAddress, "q"}
+	out, err := RunCommand("mstconfig", args...)
+	return out, err
+}
+
+func ParseMstconfigOutput(mstOutput string, attributes []string) (fwCurrent, fwNext map[string]string) {
+	glog.Infof("ParseMstconfigOutput(): Attributes %v", attributes)
+	fwCurrent = map[string]string{}
+	fwNext = map[string]string{}
+	formatRegex := regexp.MustCompile(`(?P<Attribute>\w+)\s+(?P<Default>\S+)\s+(?P<Current>\S+)\s+(?P<Next>\S+)`)
+	mstOutputLines := strings.Split(mstOutput, "\n")
+	for _, attr := range attributes {
+		for _, line := range mstOutputLines {
+			if strings.Contains(line, attr) {
+				regexResult := formatRegex.FindStringSubmatch(line)
+				fwCurrent[attr] = regexResult[3]
+				fwNext[attr] = regexResult[4]
+				break
+			}
+		}
+	}
+	return
+}
+
+func mellanoxBlueFieldMode(PciAddress string) (BlueField2Mode, error) {
+	glog.V(2).Infof("MellanoxBlueFieldMode():checking mode for card %s", PciAddress)
+	out, err := MstConfigReadData(PciAddress)
+	if err != nil {
+		glog.Errorf("MellanoxBlueFieldMode(): failed to get mlx nic fw data %v", err)
+		return -1, fmt.Errorf("failed to get mlx nic fw data %v", err)
+	}
+
+	attrs := []string{internalCPUPageSupplier, internalCPUEswitchManager, internalCPUIbVporto, internalCPUOffloadEngine}
+	mstCurrentData, _ := ParseMstconfigOutput(out, attrs)
+
+	internalCPUPageSupplierstatus, exist := mstCurrentData[internalCPUPageSupplier]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUPageSupplier)
+	}
+
+	internalCPUEswitchManagerStatus, exist := mstCurrentData[internalCPUEswitchManager]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUEswitchManager)
+	}
+
+	internalCPUIbVportoStatus, exist := mstCurrentData[internalCPUIbVporto]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUIbVporto)
+	}
+	internalCPUOffloadEngineStatus, exist := mstCurrentData[internalCPUOffloadEngine]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUOffloadEngine)
+	}
+
+	// check for DPU
```

Comment:
Just a comment: These registers aren't available with some versions of mstflint.

INTERNAL_CPU_MODEL should also be 1 as well. This is true for both DPU mode and NIC mode. Maybe we should add a check here just to be safe.

Reply:
@wizhaoredhat if for both it will be 1 why do we need to check it?

Reply:
It is possible that the NIC was incorrectly set to Separated Host Mode which will have a value of zero.

Reply:
sounds good done